src/wallet/keyman.cpp

```diff
@@ -0,0 +1,272 @@
+        }
+        return batch.WriteCryptedDescriptorKey(desc_id, pubkey, ckey);
+    } else {
+        return AddKeyInner(batch, key, pubkey) && batch.WriteDescriptorKey(desc_id, pubkey, key.GetPrivKey());
+    }
+}
+
+bool KeyManager::AddHDKey(WalletBatch& batch, const CExtKey& extkey, const CExtPubKey& extpub)
+{
+    LOCK(cs_keyman);
+    const CKeyID& id = extpub.pubkey.GetID();
+    m_map_xpubs[id] = extpub;
+    batch.WriteHDPubKey(extpub);
+
+    if (m_storage.HasEncryptionKeys()) {
+        std::vector<unsigned char> ckey = AddCryptedKeyInner(batch, extkey.key, extpub.pubkey);
+        return !ckey.empty();
+    } else {
+        return AddKeyInner(batch, extkey.key, extpub.pubkey);
+    }
+}
+
+void KeyManager::LoadKey(const CKeyID& key_id, const CKey& key)
+{
+    LOCK(cs_keyman);
+    m_map_keys[key_id] = key;
+}
+
+bool KeyManager::LoadCryptedKey(const CKeyID& key_id, const CPubKey& pubkey, const std::vector<unsigned char>& ckey)
+{
+    LOCK(cs_keyman);
+    if (!m_map_keys.empty()) {
+        return false;
+    }
+
+    m_map_crypted_keys[key_id] = make_pair(pubkey, ckey);
+    return true;
+}
+
+void KeyManager::LoadHDKey(const CKeyID& key_id, const CExtPubKey& xpub)
+{
+    LOCK(cs_keyman);
+    m_map_xpubs[key_id] = xpub;
+}
+
+bool KeyManager::CheckDecryptionKey(const CKeyingMaterial& master_key, bool accept_no_keys)
+{
+    LOCK(cs_keyman);
+    if (!m_map_keys.empty()) {
+        return false;
+    }
+
+    bool keyPass = m_map_crypted_keys.empty(); // Always pass when there are no encrypted keys
+    bool keyFail = false;
+    for (const auto& mi : m_map_crypted_keys) {
+        const CPubKey& pubkey = mi.second.first;
+        const std::vector<unsigned char>& crypted_secret = mi.second.second;
+        CKey key;
+        if (!DecryptKey(master_key, crypted_secret, pubkey, key)) {
+            keyFail = true;
+            break;
+        }
+        keyPass = true;
+        if (m_decryption_thoroughly_checked)
+            break;
+    }
+    if (keyPass && keyFail) {
+        LogPrintf("The wallet is probably corrupted: Some keys decrypt but not all.\n");
+        throw std::runtime_error("Error unlocking wallet: some keys decrypt but not all. Your wallet file may be corrupt.");
+    }
+    if (keyFail || (!keyPass && !accept_no_keys)) {
+        return false;
+    }
+    m_decryption_thoroughly_checked = true;
+    return true;
+}
+
+bool KeyManager::Encrypt(const CKeyingMaterial& master_key, WalletBatch* batch)
+{
+    LOCK(cs_keyman);
+
+    // Nothing to encrypt
+    if (m_map_keys.empty()) {
+        return true;
+    }
+
+    if (!m_map_crypted_keys.empty()) {
+        return false;
+    }
+
+    for (const auto& [id, key] : m_map_keys)
+    {
+        CPubKey pubkey = key.GetPubKey();
+        CKeyingMaterial secret(key.begin(), key.end());
+        std::vector<unsigned char> crypted_secret;
+        if (!EncryptSecret(master_key, secret, pubkey.GetHash(), crypted_secret)) {
+            return false;
```

Comment:
Clear `m_map_crypted_keys` first?

Reply:
I think it's fine to not do this. The caller will `assert(false)` in that case anyways.